process/process.go

```diff
@@ -228,15 +230,24 @@ func (p *Process) Run(ctx context.Context) error {
 		}
 
 		select {
+		case <-p.Done(): // process exited of it's own accord
+			return
 		case <-ctx.Done():
-			p.logger.Debug("[Process] Context done, terminating")
-			if err := p.Terminate(); err != nil {
-				p.logger.Debug("[Process] Failed terminate: %v", err)
+			p.logger.Debug("[Process] Context done, terminating. pid=%d", p.pid)
+			if err := p.Interrupt(); err != nil {
+				p.logger.Warn("[Process] Failed terimnation: %v", err)
 			}
-			return
 
-		case <-p.Done():
-			return
+			select {
+			case <-p.Done(): // process exited after being interrupted
+				return
+			case <-time.After(p.conf.SignalGracePeriod):
+				p.logger.Warn("[Process] Has not terminated in time, killing. pid=%d", p.pid)
+				if err := p.Terminate(); err != nil {
+					p.logger.Error("[Process] error sending SIGKILL: %s", err)
+					return
+				}
+			}
```

Comment:
This diff is the crux of the PR.

Reply:
For historical reasons, `p.Terminate` is SIGKILL, `p.Interrupt` is (by default) SIGTERM.
https://github.com/buildkite/agent/blob/5c089a8242ca77f32075229eda64396b097f7be8/process/signal.go#L35-L50